Step04/app/src/main/java/com/kwoolytech/step04/MainActivity.java

```diff
@@ -0,0 +1,168 @@
+        initializeUi();
+        initializeKwoolyHttpCallbackFunction();
+
+        queryWeatherJsonData();
+
+        return;
+    }
+
+    private void initializeUi() {
+        listViewWeatherAdapter = new HansWeatherViewAdapter(MainActivity.this, dataModel,
+                                         R.layout.listviewitem_weather);
+        ((ListView)findViewById(R.id.listViewWeather)).setAdapter(listViewWeatherAdapter);
+
+        findViewById(R.id.buttonRefresh).setOnClickListener(new Button.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                queryWeatherJsonData();
+            }
+        });
+        return;
+    }
+
+    private void initializeKwoolyHttpCallbackFunction() {
+        httpClientCallback = new KwoolyHttpCallback() {
+            @Override
+            public void onWeatherJsonDataReceived(String result) {
+                try {
+                    dataModel.setWeatherJsonData(result);
+                    queryWeatherBitmapData();
+                } catch (Exception e) {
+                    Log.e(getClass().getName(), "Exception: ");
+                    e.printStackTrace();
+                }
+            }
+
+            @Override
+            public void onWeatherBitmapDataReceived(Bitmap result) {
+                try {
+                    dataModel.setWeatherBitmapData(result);
+                    presentData();
+                } catch (Exception e) {
+                    Log.e(getClass().getName(), "Exception: ");
+                    e.printStackTrace();
+                }
+            }
+        };
+    }
+
+    private void queryWeatherJsonData() {
+        try {
+            httpClient.httpGetJsonData(
+                    CommonTool.WeatherQueryUrl + "lat=37.49&lon=127.01&units=metric" + CommonTool.ApiKey,
+                    httpClientCallback);
+        } catch (Exception e) {
+            Log.e(getClass().getName(), "Exception: ");
+            e.printStackTrace();
+        }
+        return;
+    }
+
+    private void queryWeatherBitmapData() {
+        try {
+            httpClient.httpGetBitmapData(
+                    CommonTool.WeatherIconUrl + dataModel.getIconCode() + ".png",
+                    httpClientCallback);
+        } catch (Exception e) {
+            Log.e(getClass().getName(), "Exception: ");
+            e.printStackTrace();
+        }
+        return;
+    }
+
+    private void presentData() {
+        listViewWeatherAdapter.notifyDataSetChanged();
+        return;
+    }
+
+    @Override
+    public void onBackPressed() {
+        DrawerLayout drawer = (DrawerLayout)findViewById(R.id.drawer_layout);
+        if (drawer.isDrawerOpen(GravityCompat.START)) {
+            drawer.closeDrawer(GravityCompat.START);
+        } else {
+            super.onBackPressed();
+        }
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.main, menu);
+        return true;
+    }
```

Comment:
@kwoolytech Since we are adding navigation drawer for providing menu items, this code is not necessary at the moment but let's keep it here for a later use case